### PR TITLE
Always evaluate constant operands in const-prop

### PR DESCRIPTION
**Pull request by b-naber**
r? @lcnr

### CODE REVIEW

File: src/test/ui/consts/const-eval/issue-50814.rs [hunk 1]
```diff
@@ -14,12 +14,15 @@ struct Sum<A,B>(A,B);
 impl<A: Unsigned, B: Unsigned> Unsigned for Sum<A,B> {
     const MAX: u8 = A::MAX + B::MAX;
     //~^ ERROR any use of this value will cause an error [const_err]
+    //~| ERROR any use of this value will cause an error
     //~| WARN this was previously accepted by the compiler but is being phased out
+    //~| WARNING this was previously accepted by the compiler
```
Comment (lcnr) [inline src/test/ui/consts/const-eval/issue-50814.rs:19]:
```suggestion
    //~| WARN this was previously accepted by the compiler but is being phased out
```
keep the two lines consistent

**Review by lcnr — COMMENTED** [review]
one nit, then r=me

**Comment by lcnr** [comment]
@bors try @rust-timer queue

this might actually impact perf, if unlikely

**Comment by lcnr** [comment]
the way that mostly `incr-unchanged` is impacted and all perf impacts are really small I think this isn't too relevant.

I consider this a bug fix.

@bors r+

**Comment by RalfJung** [comment]
Const-prop does not run in the body of a `const`/`static`, does it?
Those items contain promoteds that might fail to evaluate, so it is crucial that we only evaluate those nested constants that are actually needed to compute the `const`/`static` initializer. I tried to fix this but too much code out there has such potentially failing promoteds. :(  See https://github.com/rust-lang/rust/issues/80619.

I am fairly sure I added a test for this so it's probably fine, just double-checking. :)

**Comment by b-naber** [comment]
@Dylan-DPC this doesn't look spurious. Can you r- this, please?

**Comment by JohnCSimon** [comment]
Triage: 
@b-naber can you post the status of this PR?

**Comment by b-naber** [comment]
I don't know why these lints are output on that machine (which seems to be a 64-bit machine):

```
Future incompatibility report: Future breakage diagnostic:
error: any use of this value will cause an error
  --> /checkout/src/test/ui/consts/const-eval/issue-50814.rs:15:21
   |
LL |     const MAX: u8 = A::MAX + B::MAX;
   |     ----------------^^^^^^^^^^^^^^^-
   |                     |
   |                     attempt to compute `u8::MAX + u8::MAX`, which would overflow
   |
   = note: `#[deny(const_err)]` on by default
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #71800 <https://github.com/rust-lang/rust/issues/71800>

Future breakage diagnostic:
error: any use of this value will cause an error
  --> /checkout/src/test/ui/consts/const-eval/issue-50814.rs:15:21
   |
LL |     const MAX: u8 = A::MAX + B::MAX;
   |     ----------------^^^^^^^^^^^^^^^-
   |                     |
   |                     attempt to compute `u8::MAX + u8::MAX`, which would overflow
   |
   = note: `#[deny(const_err)]` on by default
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #71800 <https://github.com/rust-lang/rust/issues/71800>
```

This looks like a miri-related lint, @RalfJung do you have any idea why we get those lints on that machine, but not on the ones in the PR CI?

**Comment by RalfJung** [comment]
Maybe different flags? Debug vs release mode can make a difference.
Also the lints are emitted by const-prop-lint, which uses Miri but then also does a lot of other stuff, and I'm not much involved with that. Maybe @oli-obk knows more.

**Comment by RalfJung** [comment]
Also make sure you rebase, there have been some changes wrt binop overflow checking a few weeks ago.

File: compiler/rustc_mir_transform/src/const_prop.rs [hunk 2]
```diff
@@ -463,11 +463,6 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
 
     /// Returns the value, if any, of evaluating `c`.
     fn eval_constant(&mut self, c: &Constant<'tcx>) -> Option<OpTy<'tcx>> {
-        // FIXME we need to revisit this for #67176
-        if c.needs_subst() {
-            return None;
-        }
-
```
Comment (oli-obk) [inline compiler/rustc_mir_transform/src/const_prop.rs:470]: you need to do the same change in `const_prop_lint.rs`

**Comment by b-naber** [comment]
The error we get in the failing mir-opt test looks wrong.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -46,6 +46,9 @@ jobs:
           - name: x86_64-gnu-llvm-12
             os: ubuntu-20.04-xl
             env: {}
+          - name: dist-i586-gnu-i586-i686-musl
+            os: ubuntu-20.04-xl
+            env: {}
```
Comment (oli-obk) [inline .github/workflows/ci.yml:51]: did this accidentally get added during a rebase?

File: src/ci/github-actions/ci.yml [hunk 4]
```diff
@@ -286,7 +286,10 @@ jobs:
 
           - name: x86_64-gnu-llvm-12
             <<: *job-linux-xl
-            
+
+          - name: dist-i586-gnu-i586-i686-musl
+            <<: *job-linux-xl
+
```
Comment (oli-obk) [inline src/ci/github-actions/ci.yml:292]: same here

**Review by oli-obk — CHANGES_REQUESTED** [review]
I guess the failing mir opt test shows that we fail to add some constants to https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.Body.html#structfield.required_consts though I don't know why, how or where.

**Comment by RalfJung** [comment]
> I guess the failing mir opt test shows that we fail to add some constants to https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.Body.html#structfield.required_consts though I don't know why, how or where.

Promoteds don't get added there. That is by design since their evaluation cannot fail (in `fn`), or must only be attempted if they are in live code (in `const`/`static`).

Not sure if that's the issue here, though.

**Comment by b-naber** [comment]
I'm going to close this, not familiar enough with const prop to tell what's really going on here.

**Pull request closed without merging**